File: Python/lld-examples/easy/atm-system/transaction.py
from datetime import datetime
from enum import Enum

class TransactionType(Enum):
    """Enum representing different types of transactions"""
    DEPOSIT = "Deposit"
    WITHDRAWAL = "Withdrawal"
    BALANCE_INQUIRY = "Balance Inquiry"
    PIN_CHANGE = "PIN Change"
    TRANSFER = "Transfer"
    
    def __str__(self):
        return self.value

class Transaction:
    """
    Transaction class representing individual transactions
    """
    
    _transaction_counter = 1
    
    def __init__(self, transaction_type: TransactionType, amount: float, description: str):
        self.transaction_id = f"TXN{Transaction._transaction_counter:06d}"
        Transaction._transaction_counter += 1
        self.type = transaction_type
        self.amount = amount
        self.description = description
        self.timestamp = datetime.now()
    
    def get_formatted_timestamp(self) -> str:
        """Get formatted timestamp"""
        return self.timestamp.strftime("%Y-%m-%d %H:%M:%S")
    
    def __str__(self):
        return f"{self.transaction_id} | {self.get_formatted_timestamp()} | {self.type.value} | ${self.amount:.2f} | {self.description}" 

File: JAVA/lld-examples/easy/atm-system/src/main/java/com/systemdesign/lld/atm/Transaction.java
package com.systemdesign.lld.atm;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Transaction class representing individual transactions
 */
public class Transaction {
    private static int transactionCounter = 1;
    
    private String transactionId;
    private TransactionType type;
    private double amount;
    private String description;
    private LocalDateTime timestamp;
    
    public Transaction(TransactionType type, double amount, String description) {
        this.transactionId = "TXN" + String.format("%06d", transactionCounter++);
        this.type = type;
        this.amount = amount;
        this.description = description;
        this.timestamp = LocalDateTime.now();
    }
    
    public String getTransactionId() {
        return transactionId;
    }
    
    public TransactionType getType() {
        return type;
    }
    
    public double getAmount() {
        return amount;
    }
    
    public String getDescription() {
        return description;
    }
    
    public LocalDateTime getTimestamp() {
        return timestamp;
    }
    
    public String getFormattedTimestamp() {
        return timestamp.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
    }
    
    @Override
    public String toString() {
        return String.format("%s | %s | %s | $%.2f | %s",
                transactionId,
                getFormattedTimestamp(),
                type.getDisplayName(),
                amount,
                description);
    }
} 